datalad/interface/save.py

```diff
@@ -68,61 +98,105 @@ class Save(Interface):
+
         # shortcut
         ds = require_dataset(dataset, check_installed=True,
                              purpose='saving')
 
-        _datalad_msg = False
-        if not message:
-            message = 'Recorded existing changes'
-            _datalad_msg = True
-
+        # use the dataset's base path to indiciate that everything
+        # should be saved
         if auto_add_changes:
             files = [ds.path]
 
-        # treat special case of still untracked subdatasets.
-        # those need to become submodules now, as they are otherwise added
-        # without an entry in .gitmodules, and subsequently break Git's
-        # submodule functionality completely
-        for utf in ds.repo.repo.untracked_files:
-            utf_abspath = opj(ds.path, utf)
-            if not isdir(utf_abspath):
-                # this cannot be a repository
-                continue
-
-            # test whether the potential submodule is scheduled for saving
-            utf_realpath = realpath(utf_abspath)
-            if any([utf_realpath.startswith(_with_sep(realpath(f)))
-                    for f in files]):
-                # matches at least one path -> turn into submodule
-                _install_subds_inplace(
-                    ds=ds,
-                    path=utf_abspath,  # can be ignored, we don't need the return value
-                    relativepath=utf,
-                    name=None)
+        # track whether we modified anything, so it becomes
+        # possible to decide when/what to save further down
+        # and one level up
+        _modified_flag = False
+
+        _modified_flag = untracked_subdatasets_to_submodules(
+            ds, files)
+
+        # now we should have a complete list of submodules to potentially
+        # recurse into
+        if recursive and (recursion_limit is None or recursion_limit > 0):
+            # what subdataset to touch?
+            subdss = []
+            if auto_add_changes:
+                # all installed 1st-level ones
+                # we only want immediate subdatasets, higher depths will come via
+                # recursion
+                subdss = [Dataset(opj(ds.path, subds_path))
+                          for subds_path in ds.get_subdatasets(
+                              recursive=False)]
+            elif files is not None:
+                # only subdatasets that contain any of the to-be-considered
+                # paths
+                subdss = [ds.get_containing_subdataset(
+                    p, recursion_limit=1) for p in files]
+            # skip anything that isn't installed, or this dataset
+            subdss = [d for d in subdss if d.is_installed() and d != ds]
+
+            prop_recursion_limit = \
+                None if recursion_limit is None else max(recursion_limit - 1, 0)
+            for subds in subdss:
+                subds_modified = Save.__call__(
+                    message=message,
+                    files=files,
```

Comment:
Line 163 + line 177ff:

Why not filtering files first and pass only fitting ones to recursive calls? Right now the very same paths are checked over and over again.